cmd/maya-exporter/app/collector/jivacollector.go

```diff
@@ -0,0 +1,216 @@
+// all registered metrics. Ideally, Collector implementations support
+// concurrent readers.
+
+// Collect collects all the registered stats metrics from the Ope...nEBS volumes.
+// It tries to reconnect with the volume if there is any error via a goroutine.
+func (j *JivaStatsExporter) Collect(ch chan<- prometheus.Metric) {
+	j.collector()
+	// collect the metrics extracted by collect methods called above
+	j.Metrics.readIOPS.Collect(ch)
+	j.Metrics.readTimePS.Collect(ch)
+	j.Metrics.readBlockCountPS.Collect(ch)
+	j.Metrics.writeIOPS.Collect(ch)
+	j.Metrics.writeTimePS.Collect(ch)
+	j.Metrics.writeBlockCountPS.Collect(ch)
+	j.Metrics.actualUsed.Collect(ch)
+	j.Metrics.logicalSize.Collect(ch)
+	j.Metrics.sectorSize.Collect(ch)
+	j.Metrics.readLatency.Collect(ch)
+	j.Metrics.writeLatency.Collect(ch)
+	j.Metrics.avgReadBlockCountPS.Collect(ch)
+	j.Metrics.avgWriteBlockCountPS.Collect(ch)
+	j.Metrics.sizeOfVolume.Collect(ch)
+	j.Metrics.volumeUpTime.Collect(ch)
+	j.Metrics.connectionRetryCounter.Collect(ch)
+	j.Metrics.connectionErrorCounter.Collect(ch)
+}
+
+// collector selects the container attached storage for the collection of
+// metrics.Supported CAS are jiva and cstor.
+func (j *JivaStatsExporter) collector() error {
+	// collect the metrics from jiva controller and send it via channels
+	if err := j.collect(); err != nil {
+		j.Metrics.connectionErrorCounter.WithLabelValues(err.Error()).Inc()
+		glog.Error("Error in collecting metrics, found error:", err)
+		return errors.New("error in collecting metrics")
+	}
+	return nil
+}
+
+// NewJivaStatsExporter returns Jiva volume controller URL along with Path.
+func NewJivaStatsExporter(volumeControllerURL *url.URL) *JivaStatsExporter {
+	volumeControllerURL.Path = v1.StatsAPI
+	return &JivaStatsExporter{
+		VolumeControllerURL: volumeControllerURL.String(),
+		Metrics:             metrics,
+	}
+}
+
+// getVolumeStats is used to decode the response from the Jiva controller
+// response received by the client is in json format which then decoded
+// and mapped to VolumeMetrics.
+func (j *JivaStatsExporter) getVolumeStats(obj interface{}) error {
+	httpClient := http.DefaultClient
+	httpClient.Timeout = 1 * time.Second
+	resp, err := httpClient.Get(j.VolumeControllerURL)
+
+	if err != nil {
+		glog.Error("could not retrieve OpenEBS Volume controller metrics: %v", err)
+		return err
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(obj)
+
+	if err != nil {
+		glog.Error("could not decode OpenEBS Volume controller metrics: %v", err)
+		return err
+	}
+
+	defer resp.Body.Close()
+	return nil
+}
+
+// collect is used to set the values gathered from OpenEBS volume controller
+// to prometheus gauges.
+func (j *JivaStatsExporter) collect() error {
+	var (
+		initialMetrics, finalMetrics v1.VolumeMetrics
+		m                            MetricsDiff
+	)
+
+	err := j.getVolumeStats(&initialMetrics)
+	if err != nil {
+		glog.Error("Could not decode: %v", err)
+		return err
+	}
+
+	time.Sleep(1 * time.Second)
```

Comment:
Same comments as earlier. Can this be a property of JivaStatsExporter?
Less optimal - Put comments that metrics is collected per second in the method comments.